plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,198 @@
+                    lock_surface->output->width,
+                    lock_surface->output->height);
+                LOGI("surface_configure output=", lock_surface->output->name, " ",
+                    lock_surface->output->width, "x",
+                    lock_surface->output->height,
+                    " serial=", serial);
+
+                surfaces.push_back(lock_surface);
+
+                auto node = new std::shared_ptr<wf::scene::wlr_surface_node_t>(
+                    new lock_surface_node_t(lock_surface->surface, true));
+
+                lock_surface->data = node;
+                auto target_layer  = wf::scene::layer::TOP;
+                auto wo = wf::get_core().output_layout->find_output(lock_surface->output);
+                auto layer_node = wo->node_for_layer(target_layer);
+
+                (*node)->set_enabled(true);
+                wf::scene::add_front(layer_node, *node);
+                wf::get_core().seat->set_active_node(*node);
+
+                lock_surface_destroy.set_callback([lock_surface, this] (void *data)
+                {
+                    auto node = (std::shared_ptr<wf::scene::wlr_surface_node_t>*)lock_surface->data;
+                    if (node != nullptr)
+                    {
+                        wf::scene::remove_child(*node);
+                    }
+
+                    lock_surface_destroy.disconnect();
+                });
+                lock_surface_destroy.connect(&lock_surface->events.destroy);
+            });
+            new_surface.connect(&lock->events.new_surface);
+
+            unlock.set_callback([this] (void *data)
+            {
+                LOGI("unlock");
+            });
+            unlock.connect(&lock->events.unlock);
+
+            destroy.set_callback([this] (void *data)
+            {
+                new_surface.disconnect();
+                unlock.disconnect();
+                destroy.disconnect();
+                LOGI("destroy");
+                plugin->unlock();
+            });
+            destroy.connect(&lock->events.destroy);
+        }
+
+      private:
+        wf_session_lock_impl *plugin;
+        wlr_session_lock_v1 *lock;
+        std::vector<wlr_session_lock_surface_v1*> surfaces;
+        wf::wl_listener_wrapper new_surface;
+        wf::wl_listener_wrapper unlock;
+        wf::wl_listener_wrapper destroy;
+        wf::wl_listener_wrapper lock_surface_destroy;
+    };
+
+  public:
+    void init() override
+    {
+        auto display = wf::get_core().display;
+        manager = wlr_session_lock_manager_v1_create(display);
+
+        new_lock.set_callback([this] (void *data)
+        {
+            wlr_session_lock_v1 *wlr_lock = (wlr_session_lock_v1*)data;
+            if (cur_lock.get() == nullptr)
+            {
+                cur_lock.reset(new wayfire_session_lock(this, wlr_lock));
+                // TODO: blank the screen, or set a timeout for the lock app to commit a surface.
```

Comment:
Maybe same as in the 'crashed' case, I would create a simple cairo surface saying 'Waiting for screenlock' or 'Screenlocker crashed?' 

Here is a small example of how to show a cairo-rendered text in a node: https://github.com/WayfireWM/wayfire/blob/2e0926f8f5a1c3834c51e97673bfbf074f4506ac/plugins/single_plugins/wsets.cpp#L25-L78

If you feel like it, you could also make this into a helper class in `plugins/common/...` - a node implementation which can take cairo options and text, can be set the geometry, etc. But if you feel like this is too much, feel free to also copy-paste the code and adapt it.